### PR TITLE
[TS migration] Migrate DevTools to TS, clean other files

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details
This PR aims to migrate remaining files and clean up before migrating Onyx.js to TS.

### Related Issues
Part of https://github.com/Expensify/App/issues/34344

### Automated Tests
N/A

### Manual Tests
N/A

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/Expensify/react-native-onyx/assets/46095609/072b4c68-6e20-4b7d-9f90-28f9bd6ce168)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/46095609/6d6a03f0-cd47-4d50-8266-10a5bb922323

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/46095609/98d5e144-81a4-4e5d-a604-3fe7d173a56c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/46095609/26cde66c-4d04-4e7c-9b1c-d47c0e5565fb

https://github.com/Expensify/react-native-onyx/assets/46095609/e9a2785f-8471-41a5-bb21-270149f8eea8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/react-native-onyx/assets/46095609/9decd439-0287-4246-b2ef-5f8a157fa8c9

</details>

### CODE REVIEW

**Review by Gonals — APPROVED** [review]

File: lib/DevTools.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        try {
+            // We don't want to augment the window type in a library code, so we use type assertion instead
+            // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/no-explicit-any
+            const reduxDevtools: ReduxDevtools = (window as any).__REDUX_DEVTOOLS_EXTENSION__;
+
+            if ((options && options.remote) || typeof window === 'undefined' || !reduxDevtools) {
+                return;
+            }
+            // eslint-disable-next-line no-underscore-dangle, @typescript-eslint/no-explicit-any
+            return reduxDevtools.connect(options);
```
Comment (fabioh8010) [inline lib/DevTools.ts:48]:
```suggestion

            return reduxDevtools.connect(options);
```

I think these two rules are unnecessary, I removed here and no problems

File: lib/DevTools.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            this.defaultState = initialState;
+        } catch (e) {
+            console.error(ERROR_LABEL, e);
+        }
+    }
+
+    /**
+     * This clears the internal state of the DevTools, preserving the keys included in `keysToPreserve`
+     */
+    public clearState(keysToPreserve: string[] = []): void {
```
Comment (fabioh8010) [inline lib/DevTools.ts:93]:
```suggestion
    public clearState(keysToPreserve: string[] = []) {
```

We don't need to type `void`. Also, why `public` here? The others don't have it.

File: lib/index.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import Onyx from './Onyx';
+import type {OnyxUpdate, ConnectOptions} from './Onyx';
+import type {CustomTypeOptions, OnyxCollection, OnyxEntry, NullishDeep, KeyValueMapping, OnyxKey, Selector, WithOnyxInstanceState} from './types';
+import useOnyx from './useOnyx';
+import withOnyx from './withOnyx';
+
+export default Onyx;
+export {withOnyx, useOnyx};
+export type {CustomTypeOptions, OnyxCollection, OnyxEntry, OnyxUpdate, ConnectOptions, NullishDeep, KeyValueMapping, OnyxKey, Selector, WithOnyxInstanceState};
```
Comment (fabioh8010) [inline lib/index.ts:9]:
```suggestion
export type {CustomTypeOptions, OnyxCollection, OnyxEntry, OnyxUpdate, ConnectOptions, NullishDeep, KeyValueMapping, OnyxKey, Selector, WithOnyxInstanceState, UseOnyxResult, OnyxValue, FetchStatus};
```

Let's export these two like before, they can be useful for the library consumers

File: lib/DevTools.ts (hunk 2 above)
Comment (blazejkustra) [reply lib/DevTools.ts:93]: I think it's better to define all return types in a library such as react-native-onyx, wdyt?

File: lib/index.ts (hunk 3 above)
Comment (blazejkustra) [reply lib/index.ts:9]: Will do in another PR, thanks!

File: lib/DevTools.ts (hunk 2 above)
Comment (fabioh8010) [reply lib/DevTools.ts:93]: It works for me!

**Pull request merged**